numba/inline_closurecall.py

```diff
@@ -870,3 +872,150 @@ def fix_array_assign(stmt):
+
+        empty_func = scope.make_temp(loc)
+        fnty = get_np_ufunc_typ(np.empty)
+        sig = context.resolve_function_type(fnty, (size_typ,), {})
+        typemap[empty_func.name] = fnty #
+
+        stmts.append(_new_definition(func_ir, empty_func,
+                         ir.Global('empty', np.empty, loc=loc), loc))
+
+        empty_call = ir.Expr.call(empty_func, [size_var], {}, loc=loc)
+        calltypes[empty_call] = typing.signature(array_typ, size_typ)
+        stmts.append(_new_definition(func_ir, array_var, empty_call, loc))
+
+        for i in range(size):
+            index_var = scope.make_temp(loc)
+            index_typ = types.intp
+            typemap[index_var.name] = index_typ
+            stmts.append(_new_definition(func_ir, index_var,
+                    ir.Const(i, loc), loc))
+            setitem = ir.SetItem(array_var, index_var, seq[i], loc)
+            calltypes[setitem] = typing.signature(types.none, array_typ,
+                                                  index_typ, dtype)
+            stmts.append(setitem)
+
+        stmts.extend(dels)
+        return True
+
+    list_vars = []
+    dead_vars = []
+    list_items = []
+    stmts = []
+    dels = []
+    modified = False
+    for inst in block.body:
+        if isinstance(inst, ir.Assign):
+            if isinstance(inst.value, ir.Var):
+                if inst.value.name in list_vars:
+                    list_vars.append(inst.target.name)
+                    stmts.append(inst)
+                    continue
+            elif isinstance(inst.value, ir.Expr):
+                expr = inst.value
+                if expr.op == 'build_list':
+                    list_vars = [inst.target.name]
+                    list_items = [x.name for x in expr.items]
+                    stmts.append(inst)
+                    continue
+                elif expr.op == 'call' and expr in calltypes:
+                    arr_var = inst.target
+                    if guard(inline_array, inst.target, expr,
+                                           stmts, list_vars, dels):
+                        modified = True
+                        continue
+        elif isinstance(inst, ir.Del):
+            removed_var = inst.value
+            if removed_var in list_items:
+                dels.append(inst)
+                continue
+            elif removed_var in list_vars:
+                dead_vars.append(removed_var)
+                list_vars.remove(removed_var)
+                stmts.append(inst)
+                if list_vars == []:
+                    body = []
+                    for inst in stmts:
+                        if ((isinstance(inst, ir.Assign) and
+                             inst.target.name in dead_vars) or
+                             (isinstance(inst, ir.Del) and
+                             inst.value in dead_vars)):
+                            continue
+                        body.append(inst)
+                    stmts = body
```

Comment:
The `stmts` used here is the same `stmts` in scope. Basically we need to remove `build_list` (and all variables that are synonyms to the list variable) so that they are no longer left dangling in the code after a successful inline translation. I added some comments to explain this code.